File: timedependent.py
import numpy as np
import matplotlib.pyplot as plt
from matplotlib.animation import FuncAnimation
from typing import Callable

T = float | np.ndarray

# h_bar = 1, L = 1
class FiniteDiff:
    def __init__(self, n_x: int, n_t: int, dt: float = 1e-9) -> None:
        self.n_x = n_x
        self.n_t = n_t
        self.dx = 1/(n_x - 1)
        self.dt = dt
        self.psi: np.ndarray

    @property
    def x(self) -> np.ndarray:
        return np.linspace(0, 1, self.n_x)

    @staticmethod
    def V_gaussian(a: float = 1e4, sigma: float = 1/10, mu: float = 1/2) -> Callable[[T], T]:
        return lambda x: -a * np.exp(-(x - mu)**2/(2 * sigma**2))

    @staticmethod
    def psi0_sin(n: int = 1) -> Callable[[T], T]:
        return lambda x: 2**(1/2) * np.sin(n * np.pi * x)

    @staticmethod
    def normalize(psi: np.ndarray):
        return psi / np.sum(np.abs(psi))

    def show_potential(self):
        plt.plot(self.x, self.V_gaussian()(self.x))
        plt.show()
        plt.close()

    def fdm(self, psi0: Callable[[T], T] | None = None, V: Callable[[T], T] | None = None):
        print(self.dt/self.dx**2)
        x = self.x
        if V is None: V = self.V_gaussian(a=1e4, sigma=1/10, mu=0.5)
        if psi0 is None: psi0 = self.psi0_sin(n=1)
        potential = V(x)

        self.psi = np.zeros((self.n_t, self.n_x), dtype=np.complex_)
        self.psi[0, :] = self.normalize(psi0(x))
    
        for m, psi_m in enumerate(self.psi[:-1]):
            print(f'\r{round(100 * m/self.n_t, 2)}%', end='')
            for j, psi in enumerate(psi_m):
                psi_left = 0 if j == 0 else self.psi[m, j - 1]
                psi_right = 0 if j == self.n_x - 1 else self.psi[m, j + 1]
                self.psi[m + 1, j] = psi + 1j/2 * self.dt/self.dx**2 * (psi_left - 2 * psi + psi_right) - 1j * potential[j] * self.dt * psi
            self.psi[m + 1] = self.normalize(self.psi[m + 1])

    def animate(self, frames: int = 3000, fps: float = 30):
        x = self.x
        fig, ax = plt.subplots()
        ax.set_xlim(0, 1)
        ax.set_ylim(min(self.psi.real.min(), np.abs(self.psi).min()), max(self.psi.real.max(), np.abs(self.psi).max()))
        line_abs, = plt.plot([], [], 'r-')
        line_real, = plt.plot([], [], 'b--')

        def frame(m: int) -> None:
            print(f'\r{m}', end='')
            line_abs.set_data(x, np.abs(self.psi[m]))
            line_real.set_data(x, self.psi[m].real)
        animation = FuncAnimation(fig, func=frame, frames=range(0, self.n_t, round(self.n_t/frames)), interval=1000/fps)
        animation.save('video.mp4', fps=fps, dpi=300)


if __name__ == '__main__':
    finite = FiniteDiff(X_SAMPLES, T_SAMPLES, DELTA_T)
    finite.show_potential()
    finite.fdm()
    finite.animate(FRAMES_COUNT)

File: timedependent.cs
using System.Numerics;
using System.Globalization;

namespace Schrödinger
{
    public class TimeDependent
    {
        int xRes, timeRes;
        double dt;
        double dxSquared;
        public Complex[,] psi;

        public TimeDependent(int xRes, int timeRes, double dt)
        {
            this.xRes = xRes;
            this.timeRes = timeRes;
            this.dt = dt;
            this.dxSquared = 1.0 / ((xRes - 1.0)*(xRes - 1.0));
        }

        double[] LinSpace()
        {
            double[] x = new double[xRes];
            for (int i = 0; i < xRes; i++)
            {
                x[i] = (double)i / xRes;
            }
            return x;
        }

        static double[] V_gaussian(double[] x, double a = 1e4, double sigma = 0.1, double mu = 0.5)
        {
            double[] y = new double[x.Length];
            double shifted;
            for (int i = 0; i < x.Length; i++)
            {
                shifted = x[i] - mu;
                y[i] = -a * Math.Exp(-shifted * shifted / (2 * sigma * sigma));
            }
            return y;
        }

        static Complex[] Psi0Sin(double[] x, double n = 1)
        {
            double sum = 0;
            Complex[] y = new Complex[x.Length];
            for (int i = 0; i < x.Length; i++)
            {
                y[i] = new(Math.Sqrt(2) * Math.Sin(n * Math.PI * x[i]), 0.0);
                sum += y[i].Magnitude;
            }
            
            for (int i = 0; i < y.Length; i++) y[i] /= sum;
            
            return y;
        }

        static void Normalize(Complex[] x)
        {
            double sum = 0;
            for (int j = 0; j < x.Length; j++) sum += x[j].Magnitude;
            for (int j = 0; j < x.Length; j++) x[j] /= sum;
        }

        public void FDM(int frames)
        {
            int every = timeRes / frames;
            double[] x = LinSpace();
            double[] potential = V_gaussian(x);
            Complex[] psi0 = Psi0Sin(x);
            Console.WriteLine(dt / dxSquared);

            psi = new Complex[frames, xRes];
            
            Complex psiLeft, psiRight;
            Complex[] mCurrent = psi0;
            Complex[] mNext = new Complex[xRes];

            for (int m = 0; m < frames * every - 1; m++) {
                Console.Write("\rFDM " + (100 * (double)m/ (frames * every)).ToString("0.00"));
                for (int j = 0; j < xRes; j++)
                {
                    psiLeft = j == 0 ? Complex.Zero : mCurrent[j - 1];
                    psiRight = j == xRes - 1 ? Complex.Zero : mCurrent[j + 1];
                    mNext[j] = mCurrent[j] + Complex.ImaginaryOne / 2 * dt / dxSquared * (psiLeft - 2 * mCurrent[j] + psiRight) - Complex.ImaginaryOne * potential[j] * dt * mCurrent[j];
                    if (Complex.IsNaN(mNext[j])) Console.WriteLine("NANANANANANANANAAA HELP!!!");
                }
                Normalize(mCurrent);
                if (m % every == 0) for (int j = 0; j < xRes; j++) psi[m / every, j] = mCurrent[j];

                mCurrent = mNext;
                mNext = new Complex[xRes];
            }
        }

        public void Save(string filePath)
        {
            using (StreamWriter outfile = new StreamWriter(filePath))
            {
                outfile.WriteLine($"{psi.GetLength(0)},{psi.GetLength(1)}");
                for (int i = 0; i < psi.GetLength(0); i++)
                {
                    for (int j = 0; j < psi.GetLength(1); j++)
                    {
                        outfile.Write($"{Convert.ToString(psi[i, j].Real, CultureInfo.InvariantCulture)},{Convert.ToString(psi[i, j].Imaginary, CultureInfo.InvariantCulture)}");
                        outfile.Write(j == psi.GetLength(1) - 1 ? "\n" : ";");
                    }
                }
            }
        }
    }

    public class DoIt
    {
        static void Main(string[] args)
        {
            TimeDependent td = new(X_SAMPLES, T_SAMPLES, DELTA_T);
            td.FDM(FRAMES_NUM);
            td.Save(TO_PATH);
        }
    }
}
